# 01 Importing Libraries 

In [1]:
import pandas as pd
import numpy as np
import os

# 02 Importing Data

In [2]:
path = r'/Users/woodoooo/Desktop/Instacart Basket Analysis'
df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data' , 'Prepared Data', '4.7_ords_prods_merge_Part 1.pkl'))

In [3]:
# Change data types for multiple columns
df_ords_prods_merge = df_ords_prods_merge.astype({
    'order_id': 'int32',
    'user_id': 'int32',
    'order_number': 'int8',
    'order_day_of_week': 'int8',
    'order_hour_of_day': 'int8',
    'product_id': 'int32',
    'add_to_cart_order': 'int32',
    'reordered': 'int8',
    'days_since_prior_order': 'float32',
    'aisle_id': 'int32',
    'department_id': 'int32',
    'prices': 'float32'
    })

In [2]:
df_ords_prods_merge.head()

NameError: name 'df_ords_prods_merge' is not defined

# 03 Creating the 'Busiest days' column

### Suppose your clients have changed their minds about the labels you created in your “busiest_day” column. Now, they want “Busiest day” to become “Busiest days” (plural). This label should correspond with the two busiest days of the week as opposed to the single busiest day. At the same time, they’d also like to know the two slowest days. Create a new column for this using a suitable method.

In [6]:
df_ords_prods_merge['Busiest day'].value_counts(dropna = False)

Busiest day
Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: count, dtype: int64

In [9]:
# Count how many orders happen on each day of the week:

orders_per_day = df_ords_prods_merge['order_day_of_week'].value_counts()

In [10]:
orders_per_day

order_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [11]:
df_ords_prods_merge.loc[
    (df_ords_prods_merge['order_day_of_week'] == 0) | 
    (df_ords_prods_merge['order_day_of_week'] == 1),
    'Busiest_days'
] = 'Busiest day'

In [12]:
df_ords_prods_merge.loc[
    (df_ords_prods_merge['order_day_of_week'] == 3) | 
    (df_ords_prods_merge['order_day_of_week'] == 4),
    'Busiest_days'
] = 'Slowest day'

In [13]:
df_ords_prods_merge.loc[
    (df_ords_prods_merge['order_day_of_week'] == 6) | 
    (df_ords_prods_merge['order_day_of_week'] == 2) |
    (df_ords_prods_merge['order_day_of_week'] == 5),
    'Busiest_days'
] = 'Regularly busy day'

In [14]:
df_ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2539329        1             1                  2                  8   
2   2539329        1             1                  2                  8   
3   2539329        1             1                  2                  8   
4   2539329        1             1                  2                  8   

   days_since_prior_order  new_customer  product_id  add_to_cart_order  \
0                     NaN          True         196                  1   
1                     NaN          True       14084                  2   
2                     NaN          True       12427                  3   
3                     NaN          True       26088                  4   
4                     NaN          True       26405                  5   

   reordered                             product_name  aisle_id  \
0          0                                     Soda        77   
1          0  Organic Unsweetened Vanilla Almond Milk        91   
2          0                      Original Beef Jerky        23   
3          0               Aged White Cheddar Popcorn        23   
4          0         XL Pick-A-Size Paper Towel Rolls        54   

   department_id  prices match price_range_loc     Busiest day  \
0              7     9.0  both       Mid-range  Regularly busy   
1             16    12.5  both       Mid-range  Regularly busy   
2             19     4.4  both       Low-range  Regularly busy   
3             19     4.7  both       Low-range  Regularly busy   
4             17     1.0  both       Low-range  Regularly busy   

         Busiest_days  
0  Regularly busy day  
1  Regularly busy day  
2  Regularly busy day  
3  Regularly busy day  
4  Regularly busy day

In [15]:
df_ords_prods_merge['Busiest_days'].value_counts(dropna = False)

Busiest_days
Regularly busy day    12916111
Busiest day           11864412
Slowest day            7624336
Name: count, dtype: int64

In [18]:
df_ords_prods_merge['Busiest_days'].value_counts().sum()

np.int64(32404859)

In [20]:
df_ords_prods_merge['Busiest day'].value_counts().sum()

np.int64(32404859)

Both of the columns have the same total count of 32,404,859 rows meaning all rows have been allocated.

# 04 Creating the 'Busiest hours of the days' column

### When too many users make Instacart orders at the same time, the app freezes. The senior technical officer at Instacart wants you to identify the busiest hours of the day. Rather than by hour, they want periods of time labeled “Most orders,” “Average orders,” and “Fewest orders.” Create a new column containing these labels called “busiest_period_of_day.”

In [36]:
df_ords_prods_merge['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

In [25]:
df_ords_prods_merge.loc[df_ords_prods_merge['order_hour_of_day'].isin([9,10,11,12,13,14,15,16]), 'busiest_period_of_day'] = 'Most orders'

df_ords_prods_merge.loc[df_ords_prods_merge['order_hour_of_day'].isin([8,17,18,19]), 'busiest_period_of_day'] = 'Average orders'

df_ords_prods_merge.loc[df_ords_prods_merge['order_hour_of_day'].isin([20,21,22,23,0,1,2,3,4,5,6,7]), 'busiest_period_of_day'] = 'Fewest orders'

In [26]:
df_ords_prods_merge['busiest_period_of_day'].value_counts(dropna = False)

busiest_period_of_day
Most orders       21118071
Average orders     6700579
Fewest orders      4586209
Name: count, dtype: int64

In [27]:
df_ords_prods_merge.groupby('busiest_period_of_day')['order_hour_of_day'].unique()

busiest_period_of_day
Average orders                             [8, 19, 18, 17]
Fewest orders     [7, 20, 0, 21, 22, 5, 23, 4, 6, 1, 2, 3]
Most orders                [12, 15, 9, 14, 16, 11, 10, 13]
Name: order_hour_of_day, dtype: object

In [28]:
df_ords_prods_merge.groupby(['busiest_period_of_day', 'order_hour_of_day']).size()

busiest_period_of_day  order_hour_of_day
Average orders         8                    1718118
                       17                   2087654
                       18                   1636502
                       19                   1258305
Fewest orders          0                     218769
                       1                     115700
                       2                      69375
                       3                      51281
                       4                      53242
                       5                      87961
                       6                     290493
                       7                     891054
                       20                    976156
                       21                    795637
                       22                    634225
                       23                    402316
Most orders            9                    2454203
                       10                   2761760
                       

1. “Most orders” hours are late morning → afternoon
Hours 9–16 dominate.

10:00–15:00 are the biggest block, all above 2.6 million orders each.
Insight: This is the prime shopping window — Instacart has peak demand from late morning to afternoon.

2. “Fewest orders” happen overnight
Hours 0–6 and late night hours like 20–23 are extremely low (e.g., hour 3 only ~51k orders).
Insight: Almost no one shops between midnight and 5 AM — Instacart could schedule maintenance or run updates safely here.

3. “Average orders” fill the early morning and evening
Hours 8, 17, 18, 19 are “bridge” periods.
They have enough traffic to matter (1.2–2M orders) but aren’t overwhelming like midday.
Insight: This is where dinner-time shoppers appear — might be a good time for promotions on quick meals or ready-to-eat foods.

Overall recommendation for stakeholders
“Most orders” (9–16): Focus on app performance here — any slowdown will impact millions of users. Ads/promotions should match this heavy use period.

“Fewest orders” (night): Safest time for technical updates or app freezes for maintenance.

“Average orders” (evening): Great window for targeted marketing — people might be shopping for dinner or next day items.b

In [32]:
df_ords_prods_merge['order_hour_of_day'].count()

np.int64(32404859)

In [33]:
df_ords_prods_merge['busiest_period_of_day'].count()

np.int64(32404859)

In [34]:
df_ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2539329        1             1                  2                  8   
2   2539329        1             1                  2                  8   
3   2539329        1             1                  2                  8   
4   2539329        1             1                  2                  8   

   days_since_prior_order  new_customer  product_id  add_to_cart_order  \
0                     NaN          True         196                  1   
1                     NaN          True       14084                  2   
2                     NaN          True       12427                  3   
3                     NaN          True       26088                  4   
4                     NaN          True       26405                  5   

   reordered                             product_name  aisle_id  \
0          0                                     Soda        77   
1          0  Organic Unsweetened Vanilla Almond Milk        91   
2          0                      Original Beef Jerky        23   
3          0               Aged White Cheddar Popcorn        23   
4          0         XL Pick-A-Size Paper Towel Rolls        54   

   department_id  prices match price_range_loc     Busiest day  \
0              7     9.0  both       Mid-range  Regularly busy   
1             16    12.5  both       Mid-range  Regularly busy   
2             19     4.4  both       Low-range  Regularly busy   
3             19     4.7  both       Low-range  Regularly busy   
4             17     1.0  both       Low-range  Regularly busy   

         Busiest_days busiest_period_of_day  
0  Regularly busy day        Average orders  
1  Regularly busy day        Average orders  
2  Regularly busy day        Average orders  
3  Regularly busy day        Average orders  
4  Regularly busy day        Average orders

# 05 Exporting Data

In [37]:
# Export data to pkl
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', '4.7_ords_prods_merge_Part 2.pkl'))